File: src/lib/social-api.ts
/** API client Social Auto Publishing (TikTok / YouTube). */

import { authHeaders } from "./project-api";

export type SocialPlatform = "youtube" | "tiktok";

export interface SocialAccount {
  id: string;
  platform: SocialPlatform;
  profile_name: string;
  account_name?: string | null;
  avatar_url?: string | null;
  status: "pending" | "connected" | "expired" | "revoked";
  login_method?: string | null;
  error_message?: string | null;
  created_at: string;
}

export interface PublishJob {
  id: string;
  platform: SocialPlatform;
  title?: string | null;
  scheduled_at: string;
  status: "scheduled" | "rendering" | "uploading" | "published" | "failed" | "canceled";
  remote_url?: string | null;
  error_message?: string | null;
  clip_id?: string | null;
  project_id?: string | null;
  published_at?: string | null;
}

export interface PlatformStatus {
  siap: boolean;
  alasan?: string | null;
  login_methods?: string[];
}

export async function socialPlatforms(): Promise<Record<SocialPlatform, PlatformStatus>> {
  const res = await fetch("/api/social/platforms");
  if (!res.ok) throw new Error("Gagal memuat status platform");
  return res.json();
}

export async function socialList(): Promise<{ accounts: SocialAccount[]; jobs: PublishJob[] }> {
  const h = await authHeaders();
  const res = await fetch("/api/social/list", { headers: h });
  if (!res.ok) throw new Error("Gagal memuat akun sosial");
  return res.json();
}

export async function socialConnect(
  platform: SocialPlatform,
  profileName: string,
  loginMethod?: string,
): Promise<{ auth_url: string; account_id: string }> {
  const h = await authHeaders();
  const res = await fetch("/api/social/connect", {
    method: "POST",
    headers: h,
    body: JSON.stringify({
      platform,
      profile_name: profileName,
      ...(loginMethod ? { login_method: loginMethod } : {}),
    }),
  });
  if (!res.ok) {
    const d = (await res.json().catch(() => ({}))) as { detail?: string };
    throw new Error(d.detail ?? "Gagal memulai koneksi");
  }
  return res.json();
}

export async function socialDisconnect(accountId: string): Promise<void> {
  const h = await authHeaders();
  const res = await fetch(`/api/social/accounts/${accountId}`, { method: "DELETE", headers: h });
  if (!res.ok) throw new Error("Gagal memutuskan akun");
}

export interface SocialProject {
  id: string;
  title: string;
  status: string;
  created_at: string;
}
export interface SocialClip {
  id: string;
  project_id: string;
  title?: string | null;
  start_time: number;
  end_time: number;
  rendered_url?: string | null;
}

export async function socialClips(): Promise<{ projects: SocialProject[]; clips: SocialClip[] }> {
  const h = await authHeaders();
  const res = await fetch("/api/social/clips", { headers: h });
  if (!res.ok) throw new Error("Gagal memuat proyek");
  return res.json();
}

export async function socialSchedule(
  accountIds: string[],
  clipIds: string[],
  hours?: number[],
): Promise<{ dibuat: number; jobs: PublishJob[] }> {
  const h = await authHeaders();
  const res = await fetch("/api/social/schedule", {
    method: "POST",
    headers: h,
    body: JSON.stringify({
      account_ids: accountIds,
      clip_ids: clipIds,
      ...(hours && hours.length > 0 ? { hours } : {}),
    }),
  });
  if (!res.ok) {
    const d = (await res.json().catch(() => ({}))) as { detail?: string };
    throw new Error(d.detail ?? "Gagal menjadwalkan");
  }
  return res.json();
}

export async function socialCancel(jobId: string): Promise<void> {
  const h = await authHeaders();
  const res = await fetch(`/api/social/jobs/${jobId}/cancel`, { method: "POST", headers: h });
  if (!res.ok) {
    const d = (await res.json().catch(() => ({}))) as { detail?: string };
    throw new Error(d.detail ?? "Gagal membatalkan");
  }
}

File: backend/app/social_api.py
"""Endpoint SOCIAL AUTO PUBLISHING — didaftarkan dari main.py.

Dipisah supaya main.py tidak membengkak, pola yang sama dengan broll_api.py.
"""
from __future__ import annotations

from typing import Any, Optional

from fastapi import Header, HTTPException, Request
from fastapi.responses import HTMLResponse
from pydantic import BaseModel


class ConnectIn(BaseModel):
    platform: str
    profile_name: str
    login_method: Optional[str] = None


class ScheduleIn(BaseModel):
    account_ids: list[str]
    clip_ids: list[str]
    hours: Optional[list[int]] = None


def _halaman_tutup(judul: str, pesan: str, ok: bool) -> HTMLResponse:
    """Halaman kecil untuk tab OAuth: memberi tahu hasil lalu menutup diri.

    Dibuat sebagai HTML polos (bukan redirect ke SPA) supaya pengguna melihat
    hasilnya walau tab dibuka di browser lain / mode privat, dan supaya
    halaman utama bisa memuat ulang daftar akunnya sendiri.
    """
    warna = "#16a34a" if ok else "#dc2626"
    return HTMLResponse(f"""<!doctype html><html lang="id"><head>
<meta charset="utf-8"><meta name="viewport" content="width=device-width,initial-scale=1">
<title>{judul}</title>
<style>
  :root {{ color-scheme: light dark; }}
  body {{ margin:0; min-height:100vh; display:grid; place-items:center;
         font:15px/1.6 system-ui,-apple-system,"Segoe UI",sans-serif;
         background:#0f1115; color:#f4f4f5; padding:24px; }}
  .kotak {{ max-width:420px; text-align:center; }}
  h1 {{ font-size:20px; margin:0 0 10px; color:{warna}; }}
  p {{ margin:0 0 18px; color:#a1a1aa; }}
  button {{ font:inherit; padding:10px 18px; border-radius:10px; border:0;
           background:#f4f4f5; color:#0f1115; font-weight:600; cursor:pointer; }}
</style></head><body><div class="kotak">
<h1>{judul}</h1><p>{pesan}</p>
<button onclick="tutup()">Tutup jendela ini</button>
<script>
  try {{ if (window.opener) window.opener.postMessage(
        {{ type: "cortexclip:social", ok: {str(ok).lower()} }}, "*"); }} catch (e) {{}}
  function tutup() {{ window.close(); location.href = "/social"; }}
  setTimeout(tutup, 2600);
</script></div></body></html>""")


def register_social_routes(app, get_user) -> None:
    """Daftarkan endpoint auto-publish. `get_user` diambil dari main.py."""

    @app.get("/api/social/platforms")
    async def api_social_platforms():
        """Platform mana yang siap dipakai (kredensial OAuth ada di server)."""
        from .social_publish import status_platform
        return status_platform()

    @app.get("/api/social/list")
    async def api_social_list(request: Request,
                              authorization: Optional[str] = Header(None)):
        user = await get_user(request, authorization)
        from .social_schedule import daftar
        try:
            return await daftar(str(user["id"]))
        except Exception as exc:
            raise HTTPException(400, str(exc)[:200])

    @app.post("/api/social/connect")
    async def api_social_connect(body: ConnectIn, request: Request,
                                 authorization: Optional[str] = Header(None)):
        """Mulai OAuth: balikin auth_url yang harus dibuka pengguna."""
        user = await get_user(request, authorization)
        from .social_publish import mulai_koneksi
        try:
            return await mulai_koneksi(str(user["id"]), body.platform,
                                       body.profile_name, body.login_method)
        except PermissionError as exc:
            # kredensial belum diatur admin — JUJUR, bukan diam-diam gagal
            raise HTTPException(503, str(exc)[:300])
        except ValueError as exc:
            raise HTTPException(400, str(exc)[:200])
        except Exception as exc:
            raise HTTPException(500, str(exc)[:200])

    @app.get("/api/social/callback/{platform}")
    async def api_social_callback(platform: str, request: Request):
        """Callback OAuth. TIDAK butuh header Authorization (browser redirect);
        keamanannya dari `state` acak yang mengikat baris pending."""
        code = request.query_params.get("code")
        state = request.query_params.get("state")
        err = request.query_params.get("error")
        if err:
            return _halaman_tutup("Koneksi dibatalkan",
                                  f"Platform menolak: {err}", False)
        if not code or not state:
            return _halaman_tutup("Koneksi gagal",
                                  "Callback tanpa code/state.", False)
        from .social_publish import selesaikan_koneksi
        try:
            hasil = await selesaikan_koneksi(platform, code, state)
            acc = hasil.get("account") or {}
            nama = acc.get("account_name") or acc.get("profile_name") or "akun"
            if hasil.get("nama_cocok"):
                return _halaman_tutup("Akun tersambung",
                                      f"{nama} siap dipakai auto-publish.", True)
            return _halaman_tutup(
                "Tersambung (periksa akun)",
                f"Tersambung sebagai {nama} — berbeda dari nama yang kamu tulis. "
                "Kalau salah akun, putuskan lalu sambungkan ulang.", True)
        except Exception as exc:
            return _halaman_tutup("Koneksi gagal", str(exc)[:200], False)

    @app.delete("/api/social/accounts/{account_id}")
    async def api_social_disconnect(account_id: str, request: Request,
                                    authorization: Optional[str] = Header(None)):
        user = await get_user(request, authorization)
        from .social_schedule import putuskan
        try:
            return await putuskan(str(user["id"]), account_id)
        except ValueError as exc:
            raise HTTPException(404, str(exc)[:200])
        except Exception as exc:
            raise HTTPException(400, str(exc)[:200])

    @app.get("/api/social/clips")
    async def api_social_clips(request: Request,
                               authorization: Optional[str] = Header(None)):
        """Proyek + klip milik pengguna, untuk layar 'pilih proyek'."""
        user = await get_user(request, authorization)
        from .social_publish import _sb
        try:
            proyek = await _sb("GET", "projects?select=id,title,status,created_at"
                                      f"&user_id=eq.{user['id']}"
                                      "&order=created_at.desc&limit=60")
            ids = [str(p["id"]) for p in (proyek or [])]
            klip: list[dict[str, Any]] = []
            if ids:
                klip = await _sb("GET", "clips?select=id,project_id,title,"
                                        "start_time,end_time,rendered_url"
                                        "&project_id=in.(" + ",".join(ids) + ")"
                                        "&order=start_time.asc&limit=400") or []
            return {"projects": proyek or [], "clips": klip}
        except Exception as exc:
            raise HTTPException(400, str(exc)[:200])

    @app.post("/api/social/schedule")
    async def api_social_schedule(body: ScheduleIn, request: Request,
                                  authorization: Optional[str] = Header(None)):
        user = await get_user(request, authorization)
        from .social_schedule import jadwalkan_klip
        try:
            return await jadwalkan_klip(str(user["id"]), body.account_ids,
                                        body.clip_ids, body.hours)
        except ValueError as exc:
            raise HTTPException(400, str(exc)[:250])
        except Exception as exc:
            raise HTTPException(500, str(exc)[:250])

    @app.post("/api/social/jobs/{job_id}/cancel")
    async def api_social_cancel(job_id: str, request: Request,
                                authorization: Optional[str] = Header(None)):
        user = await get_user(request, authorization)
        from .social_schedule import batalkan
        try:
            return await batalkan(str(user["id"]), job_id)
        except ValueError as exc:
            raise HTTPException(400, str(exc)[:200])
        except Exception as exc:
            raise HTTPException(500, str(exc)[:200])
